### ISSUE TITLE
The state of the "Need work" toggle button is not reported through accessibility APIs

### ISSUE THREAD

**Issue by bdorer** [vslavik/poedit#693]
This Checkbox is announced as a button using NVDA and Windows Narator. Buttons are pressed typically with enter which isn't working. Pressing Space to switch works like a checkbox. It would be great if you'll make sure that switching is announced by nvda or narator.

**Comment by vslavik** [comment]
It's not a checkbox, it's a toggle — and after checking wxWidgets' source code, is using the appropriate `BS_AUTOCHECKBOX` and `BS_PUSHLIKE` styles.

So it's not very clear to me where do you see the incorrectness?

**Comment by pvagner** [comment]
This is difficult to fix with wxToggleButton I am afraid. It's why @bdorer is asking to consider changing to checkbox instead.
Toggle button should have at least two states it's either pressed or it's not pressed.
WxWidgets implements MSAA in order to signal control roles, states and other properties to apps such as assistive technologies.
And in this particular case STATE_SYSTEM_PRESSED is never set on this control. In contrast even wx checkboxes do communicate their checked state to MSAA on windows.

**Comment by vslavik** [comment]
wxWidgets doesn't implement MSAA here; the native win32 control does. And that is the same BUTTON class for either of the two controls, so it really is something that *should* in principle be doable.

However,  your reply seems orthogonal to the original report, and speaks in completely different terms, without explaining the vagueness of the OP - the enter vs space thing in particular...

**Comment by CyrilleB79** [comment]
Additional note:
I have made a test GUI in wxPython with a wx.ToggleButton. NVDA detects it with the IAccessible accRole ROLE_SYSTEM_CHECKBUTTON whereas Poedit's one has the accRole ROLE_SYSTEM_PUSHBUTTON.
And wxPython's one it has the IAccessible accState STATE_SYSTEM_CHECKED when it is pushed.

Visually however, it does not look like Poedit's one. wx.Python's one looks like a rectangular button that can be pushed or not, whereas Poedit's one looks like a short stick with round ends and big dot left or right to denote the active (a.k.a. pressed) or unactive state. But maybe it's just design.

**Comment by bdorer** [comment]
For me it isn't important wheather it is a toggle button, a checkbox or s.th. like that. I simply was confused to press space to activate a button.

**Comment by CyrilleB79** [comment]
There are two issues described in this ticket:
1. The "Need work" push button state can be modified only pressing space, not Enter.
2. The state (pressed / not pressed ) of the "Need work" button is not reported via accessibility APIs.

For clarity, I have opened #721 to discuss this second point.
From now, let's discuss only the first point in this ticket, i.e. the way the "Need work" push button should be activated.

**Comment by vslavik** [comment]
> I have made a test GUI in wxPython with a wx.ToggleButton. NVDA detects it with the IAccessible accRole ROLE_SYSTEM_CHECKBUTTON whereas Poedit's one has the accRole ROLE_SYSTEM_PUSHBUTTON.
And wxPython's one it has the IAccessible accState STATE_SYSTEM_CHECKED when it is pushed.

Seems you're *this* close to identifying where the problem occurs, i.e. what the difference between your secret test code and publicly accessible Poedit's code is…

> But maybe it's just design.

It is. The code is public, you can see it yourself that Poedit overrides just the painting.

> For clarity, I have opened #721 to discuss this second point.

Clarity-wise, that was a bit unfortunate. As previously explained, and as you observed, native control is used - the differences in (a11y) appearance and control are evident consequences of the underlying difference in control's setup. The culprit is not immediately evident to me yet, will need some debugging.

**Comment by leonardder** [comment]
I somehow missed this issue while investigating issues for the accessibility label, I'm sorry for that.

I opened #804 and also shared my theory in the pr description about why this doesn't work out of the box.
